### ISSUE TITLE
[HOLD for payment 2023-01-05] [$4000] Emoji Picker list jumps while scrolling through arrow keys reported by @parasharrajat 

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#12772]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Login to the app on web.
2. Navigate to any report.
3. Open the Emoji Picker.
4. User down arrow key to scroll the list until you reach the section Travel & Places.
5. Now keep on scrolling down via arrow key .
6. Notice that list jumps and highlighted emoji is out of focus.

## Expected result: 
Highlighted emoji should remain in focus and list does not jump positions.

## Actual: 
Emoji List jump positions and highlighted emoji goes out of focus.

## Workaround:
unknown

## Platform:

Where is this issue occurring?

- Web
- Desktop App

**Version Number:**  1.2.28-0
**Reproducible in staging?:** y
**Reproducible in production?:** y
**Email or phone of affected tester (no customers):** 
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 

https://user-images.githubusercontent.com/43996225/202214414-eb1f460c-a2bb-4a90-b5fb-9eeac0a76383.mp4

https://user-images.githubusercontent.com/43996225/202214442-562dd0c3-0523-4b0f-94a6-acbc040e3391.mp4

**Expensify/Expensify Issue URL:** 
**Issue reported by:** @parasharrajat 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1668550863631129

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~011a004a1208ec7fec</li>
        <li>Upwork Job ID: 1598460283393093632</li>
        <li>Last Price Increase: 2022-12-15</li>
    </ul>
</details>

**Comment by JmillsExpensify** [comment]
Taking this one given that I'm on the emoji tracking issue. Not treating this as an accessibility issue.

**Comment by JmillsExpensify** [comment]
I'm wavering a bit based on my last comment. This bug seems to only affect that one section right? If this is a clear and quick fix, then I'm more amenable to addressing it now, even though we're not focused on keyboard/accessibility at the moment. Getting another opinion from engineering.

**Comment by JmillsExpensify** [comment]
@Gonals Please see above for a second opinion. Thanks!

**Comment by parasharrajat** [comment]
This issue does not come under A11y/navigation issues. This is a bug with emoji list viewing. This does not affect only one section. It might affect more sections we add them. It is just a coincidence that the section lies where the bug occurs. The main issue is with the Flat list scrolling/rendering. Due to virtual scrolling, something bad is happening with our scrolling logic.

**Comment by JmillsExpensify** [comment]
Ok well I think that's also my point here. Is this the only place this issue is happening with FlatList. I'm less and less interested is solving issues like this one as if this is the only place this is happening, and I think whatever solution we come up with, let's make sure that we aren't working on this same thing somewhere else in the repo in a silo.

**Comment by Gonals** [comment]
> @Gonals Please see above for a second opinion. Thanks!

It took me quite a bit to actually `see` the issue 😅.

If the behavior is only happening in this specific section of the emoji list, I honestly don't think it is a bit deal. If we have a quick fix, great, otherwise... 🤷.

If we start seeing scrolling issues with our flatlists elsewhere, then we'd certainly need to tackle it with more urgency

**Comment by parasharrajat** [comment]
Why do we think that this is not a  big deal? IMO, It is affecting user experience, and thus should be fixed.

**Comment by JmillsExpensify** [comment]
Still feel the same as last week, though open to differing opinions.

**Comment by parasharrajat** [comment]
Make sense https://github.com/Expensify/App/issues/12772#issuecomment-1320579574 :+1:

**Comment by JmillsExpensify** [comment]
Given that we've decided that WishList is quite far away, I'm open to re-evaluating this one. @stitesExpensify curious to get your take on this issue since you're also working on the larger emoji initiative. Maybe we even take this internal via either you or @Gonals to see if a quick fix exists?

**Comment by stitesExpensify** [comment]
I am not able to reproduce this issue:

https://user-images.githubusercontent.com/42391420/205135213-e2ace602-ee8e-448f-bf69-3eb17927e79a.mp4

**Comment by parasharrajat** [comment]
I think you have to scroll a little further in the travel & places section

**Comment by stitesExpensify** [comment]
🤦 yep you're right

**Comment by JmillsExpensify** [comment]
Yeah it's real tough to reproduce. It's a bug, though I do think it's pretty low value. Removing the FlatList hold, though I concern I have is whether the solution is straightforward around the scrolling logic. @stitesExpensify @Gonals shall we open up to External in any case?

**Comment by stitesExpensify** [comment]
Let me look into it a little deeper and see if it's easy. If I don't figure it out quickly I'll open it up

**Comment by stitesExpensify** [comment]
I think that the problem is in [this method](https://github.com/Expensify/App/blob/main/src/components/EmojiPicker/EmojiPickerMenu/index.js#L328), but i'm not positive. I'm going to make this external since I've already got 2 bugs I'm working on personally.

**Comment by stitesExpensify** [comment]
Sorry @shawnborton misclick. Unassigning you

**Comment by Santhosh-Sellavel** [comment]
Looking for proposals

**Comment by JmillsExpensify** [comment]
Doubling the price of this issue to try and get more interest.

**Comment by Pujan92** [comment]
**Proposal**

Issue: I am seeing that on scroll the `FlatList` initially removes the items from the dom and applies that height to the `div` and after some more scrolling, it removes the height also from the `div` which seems to be causing this inconsistent scroll issue. This video helps to understand.
https://user-images.githubusercontent.com/14358475/206183860-8ba79691-04e7-477b-a188-d530efc69b0c.mp4

To solve the specific issue, we can pass the max size to `maxToRenderPerBatch` within `VirtualizedList/FlatList` which won't take advantage of the virtualization and won't remove the div from the dom.
https://github.com/Expensify/App/blob/cb9a54138bafac9d39598e492669cd01a9b31af5/src/components/EmojiPicker/EmojiPickerMenu/index.js#L496
`maxToRenderPerBatch={this.emojis.length}`

https://user-images.githubusercontent.com/14358475/206185097-63693ffd-d46e-4354-8e0e-4637bb69fefd.mp4

**Comment by Santhosh-Sellavel** [comment]
@Ollyws 
Your `getItemLayout` solution not working as expected when using an upward Arrow after scrolling down.

`windowSize` seems to work well, not sure about the impact we might have on performance/storage based on the con [here](https://reactnative.dev/docs/optimizing-flatlist-configuration#windowsize:~:text=Cons%3A%20For%20a%20bigger%20windowSize%2C%20you%20will%20have%20more%20memory%20consumption.%20For%20a%20lower%20windowSize%2C%20you%20will%20have%20a%20bigger%20chance%20of%20seeing%20blank%20areas.) @Gonals Any thought on this?
 
@Pujan92 
> To solve the specific issue, we can pass the max size to maxToRenderPerBatch within VirtualizedList/FlatList which won't take advantage of the virtualization and won't remove the div from the dom.

We don't want to disable virtualization here.

So no.


Also I am also thinking solution should be fix this method mostly [this method](https://github.com/Expensify/App/blob/main/src/components/EmojiPicker/EmojiPickerMenu/index.js#L328) but sure either. B

**Comment by Pujan92** [comment]
I think for the `getItemLayout` all the items need to be of the same size to make it work, how about making the header height `32px` too?


https://github.com/Expensify/App/blob/699d46986253282403ecb4a0b74a82643a57c8be/src/styles/styles.js#L1437
Update to
```
height: 32,
display: 'flex',
alignItems: 'center',
```

**Comment by Santhosh-Sellavel** [comment]
No, we won't be doing that mostly as it is a design requirement! cc: @Gonals

**Comment by JmillsExpensify** [comment]
Thanks for the great discussion! Still working through proposals.

**Comment by priyeshshah11** [comment]
## Proposal

## Solution 1

We can add the `getItemLayout` prop to correctly calculate the offset by calculating the exact height of the elements that are before the current index. 

This solution also takes care of this issue

> Sometimes it partially is not visible. Look at this case, scroll to the end of the list, and use the up arrow mostly it's hidden.

```
getItemLayout={(list, index) => {
  const passedHeaders = _.filter(this.state.headerIndices, i => i < index).length;
  const passedNonHeaderRows = index - passedHeaders;
  const offset = (CONST.EMOJI_PICKER_ITEM_HEIGHT * passedNonHeaderRows) + (passedHeaders * CONST.EMOJI_PICKER_HEADER_HEIGHT);
  return (
      {length: CONST.EMOJI_PICKER_ITEM_HEIGHT, offset, index}
  );
}}
```


## Solution 2

We just add the `initialNumToRender={this.state.filteredEmojis.length}` to the flatlist, this fixes the issue as it does not unmount any items even after we have gone past the windowSize, thus our contentSize & offset don't change unexpectedly.

Con: Not super efficient to render all items at once.

## Solution 3

Combine both solution 1 & 2 to be at least somewhat efficient than solution 2 by providing `getItemLayout`

@Santhosh-Sellavel @Gonals @JmillsExpensify

**Comment by Pujan92** [comment]
> ## Solution 1
> We can add the `getItemLayout` prop to correctly calculate the offset by calculating the exact height of the elements that are before the current index.
> 
> ```
> getItemLayout={(list, index) => {
>   const passedHeaders = _.filter(this.state.headerIndices, i => i < index).length;
>   const passedNonHeaderRows = index - passedHeaders;
>   const offset = (CONST.EMOJI_PICKER_ITEM_HEIGHT * passedNonHeaderRows) + (passedHeaders * CONST.EMOJI_PICKER_HEADER_HEIGHT);
>   return (
>       {length: CONST.EMOJI_PICKER_ITEM_HEIGHT, offset, index}
>   );
> }}
> ```

with this, the issue `partially is not visible` seems to still persist.

<img width="347" alt="Screenshot 2022-12-08 at 7 13 47 PM" src="https://user-images.githubusercontent.com/14358475/206462290-b74f4421-346b-482e-a792-8b19ebadfb1a.png">

**Comment by priyeshshah11** [comment]
> with this, the issue `partially is not visible` seems to still persist.
>

Attached the video above, I cannot seem to reproduce

**Comment by Pujan92** [comment]
> @Pujan92 Attached the video above, I cannot seem to reproduce

Plz try to go top from bottom, you will see that for multiple sections it happens.

**Comment by priyeshshah11** [comment]
> > @Pujan92 Attached the video above, I cannot seem to reproduce
> 
> Plz try to go top from bottom, you will see that for multiple sections it happens.

Ohh yes found it, that's because the CONST.EMOJI_PICKER_HEADER_HEIGHT is incorrect, in the code it is defined as 38 but in reality it is 37, so we should change that as that's causing incorrect calculations in `getItemLayout`

<img width="366" alt="Screen Shot 2022-12-09 at 1 09 41 AM" src="https://user-images.githubusercontent.com/38547776/206467365-2424b6f5-9581-46d9-9d4c-b386d0dbcae9.png">

**Comment by Pujan92** [comment]
> Ohh yes found it, that's because the CONST.EMOJI_PICKER_HEADER_HEIGHT value is incorrect, in the code it is defined as 38 but in reality it is 37, so we should change that as that's causing incorrect calculations in `getItemLayout`

Yes, now while going bottom the issue occurs after 'Travel & Places' section. As per my understanding, the issue exists when FlatList starts creating dynamic height div onscroll.

**Comment by JmillsExpensify** [comment]
@Santhosh-Sellavel Thoughts on the more recent convo?

**Comment by priyeshshah11** [comment]
## Proposal 4 - Further Optimisations

@Santhosh-Sellavel @JmillsExpensify This fixes all the above mentioned bugs & adds performance improvements.

### Solution 

We calculate the row sizes for the full list only once at the start of the component & cache that in a map on which we can do lookups by row index in `getItemLayout`. Then we just return the normal `getItemLayout` values when the list is in search mode as we do not show the headers while searching which makes all row items the same height so we don't have to do any complex computations.

**Add this in the constructor before `this.currentScrollOffset = 0;`**

```
this.getEmojiListItemLayout = this.getEmojiListItemLayout.bind(this);
this.rowSizes = this.calculateRowSizes(this.emojis);
```

**Add this after `setupEventHandlers`**
```
/**
 * Returns the length, offset and index for a renderItem row.
 * Uses the cached size if the list is not in search mode.
 *
 * @param {Array} data
 * @param {Number} index
 *
 * @return {Object}
 */
getEmojiListItemLayout(data, index) {
    if (this.state.filteredEmojis.length !== this.emojis.length) {
        // just return the item height because headers are not shown during search
        return {
            length: CONST.EMOJI_PICKER_ITEM_HEIGHT,
            offset: CONST.EMOJI_PICKER_ITEM_HEIGHT * index,
            index,
        };
    }
    return this.rowSizes[index];
}

/**
 * Calculates all emoji row sizes and returns a map of index to row sizes (length, offset & index).
 *
 * @param {Array} emojiList
 * @return {Object}
 */
calculateRowSizes(emojiList) {
    let offset = 0;
    const sizesByIndex = {};
    _.each(emojiList, (emoji, emojiIndex) => {
        const columnIndex = emojiIndex % CONST.EMOJI_NUM_PER_ROW;

        // Calculate size only once for each row so ignore every column after first emoji in the row.
        if (columnIndex > 0) { return; }
        const index = (emojiIndex - columnIndex) / CONST.EMOJI_NUM_PER_ROW;
        const length = emoji.header ? CONST.EMOJI_PICKER_HEADER_HEIGHT : CONST.EMOJI_PICKER_ITEM_HEIGHT;
        sizesByIndex[index] = {length, offset, index};
        offset += length;
    });
    return sizesByIndex;
}

```

**Add this in the FlatList props `getItemLayout={this.getEmojiListItemLayout}`**

Also I don't think we need to change the header size to 37 anymore as it now comes up as 38, tested with the above changes.

**Comment by Santhosh-Sellavel** [comment]
@JmillsExpensify Will review this tonight, but both look little over-engineered

**Comment by s77rt** [comment]
### Proposal
```diff
diff --git a/src/CONST.js b/src/CONST.js
index 935ce523f7..809c1ef3d7 100755
--- a/src/CONST.js
+++ b/src/CONST.js
@@ -497,7 +497,7 @@ const CONST = {
     },
     NON_NATIVE_EMOJI_PICKER_LIST_HEIGHT: 298,
     EMOJI_PICKER_ITEM_HEIGHT: 32,
-    EMOJI_PICKER_HEADER_HEIGHT: 38,
+    EMOJI_PICKER_HEADER_HEIGHT: 37,
     COMPOSER_MAX_HEIGHT: 125,
     CHAT_FOOTER_MIN_HEIGHT: 65,
     CHAT_SKELETON_VIEW: {
diff --git a/src/components/EmojiPicker/EmojiPickerMenu/index.js b/src/components/EmojiPicker/EmojiPickerMenu/index.js
index 2d97a24439..6073e17aab 100755
--- a/src/components/EmojiPicker/EmojiPickerMenu/index.js
+++ b/src/components/EmojiPicker/EmojiPickerMenu/index.js
@@ -333,21 +333,19 @@ class EmojiPickerMenu extends Component {
         }
 
         // Calculate the scroll offset at the bottom of the currently highlighted emoji
-        // (subtract numHeaders because the highlightedIndex includes them, and add 1 to include the current row)
-        const numEmojiRows = (Math.floor(this.state.highlightedIndex / this.numColumns) - numHeaders) + 1;
+        const numEmojiRows = Math.floor(this.state.highlightedIndex / this.numColumns) - numHeaders;
 
         // The scroll offsets at the top and bottom of the highlighted emoji
-        const offsetAtEmojiBottom = ((numHeaders) * CONST.EMOJI_PICKER_HEADER_HEIGHT)
+        const offsetAtEmojiTop = ((numHeaders) * CONST.EMOJI_PICKER_HEADER_HEIGHT)
             + (numEmojiRows * CONST.EMOJI_PICKER_ITEM_HEIGHT);
-        const offsetAtEmojiTop = offsetAtEmojiBottom - CONST.EMOJI_PICKER_ITEM_HEIGHT;
+        const offsetAtEmojiBottom = offsetAtEmojiTop + CONST.EMOJI_PICKER_ITEM_HEIGHT;
 
         // Scroll to fit the entire highlighted emoji into the window if we need to
         let targetOffset = this.currentScrollOffset;
         if (offsetAtEmojiBottom - this.currentScrollOffset >= CONST.NON_NATIVE_EMOJI_PICKER_LIST_HEIGHT) {
             targetOffset = offsetAtEmojiBottom - CONST.NON_NATIVE_EMOJI_PICKER_LIST_HEIGHT;
-        } else if (offsetAtEmojiTop - CONST.EMOJI_PICKER_HEADER_HEIGHT <= this.currentScrollOffset) {
-            // There is always a sticky header on the top, subtract the EMOJI_PICKER_HEADER_HEIGHT from offsetAtEmojiTop to get the correct scroll position.
-            targetOffset = offsetAtEmojiTop - CONST.EMOJI_PICKER_HEADER_HEIGHT;
+        } else if (offsetAtEmojiTop <= this.currentScrollOffset) {
+            targetOffset = offsetAtEmojiTop;
         }
         if (targetOffset !== this.currentScrollOffset) {
             // Disable pointer events so that onHover doesn't get triggered when the items move while we're scrolling
@@ -506,7 +504,6 @@ class EmojiPickerMenu extends Component {
                             extraData={
                               [this.state.filteredEmojis, this.state.highlightedIndex, this.props.preferredSkinTone]
                             }
-                            stickyHeaderIndices={this.state.headerIndices}
                             onScroll={e => this.currentScrollOffset = e.nativeEvent.contentOffset.y}
                         />
                     )}
```

### Details
The root issue is caused by the use of sticky headers and the incorrect `EMOJI_PICKER_HEADER_HEIGHT` value

- `EMOJI_PICKER_HEADER_HEIGHT`
    Before the use of `ExpensifyNeue-Regular` font the height was actually 38. But now it's 37.
- Sticky headers
    Sticky headers make scrolling buggy.
    In my proposal, I removed the use of sticky headers, so I'm not actually fixing the root issue but rather avoiding it. But still I think this is an acceptable approach in our case and here is why:
    1. Sticky headers is a feature that nobody asked for. I have checked few social networks and they don't have sticky headers on their emoji picker, so we are not missing a feature that a user may be expecting.
    2. No sticky headers means more space for the emoji, and from design/ux perspective this may be actually better
    3. This is the cleanest solution so far

**Comment by bernhardoj** [comment]
@s77rt Actually, removing the `stickyHeaderIndices` is enough solving this GH issue. But I'm not sure they want to remove a feature because of a bug.

**Comment by JmillsExpensify** [comment]
@Santhosh-Sellavel @Gonals I wanted to give you both a heads up that I'm OOO this week. I intend to check in every couple of days, so if you need action on something faster, please either take it on or float the task in #expensify-open-source or #bug-zero. Thanks! 🙌🏼

**Comment by Santhosh-Sellavel** [comment]
@Gonals 

I Like @Ollyws proposal [here](https://github.com/Expensify/App/issues/12772#issuecomment-1343344297) works well for me in all cases, which I feel is the first complete proposal here, all other proposal looks similar.

We have a different approach from @s77rt suggesting removing the sticky header [here](https://github.com/Expensify/App/issues/12772#issuecomment-1345260978). I'm not sure we would do that, so let's confirm that with @Expensify/design here

**Comment by shawnborton** [comment]
I think we should keep the sticky headers personally.

**Comment by priyeshshah11** [comment]
@Santhosh-Sellavel @Gonals Does this solution not work? And isn't it optimised and better for performance than the other proposal?

**Comment by Santhosh-Sellavel** [comment]
Yes, @priyeshshah11!

**Comment by Pujan92** [comment]
> I think we should keep the sticky headers personally.

Do we need to keep the height of the header to `37px` only or are we ok to change it to `32px`?

**Comment by shawnborton** [comment]
I think so, can you share screenshots though? Also, why is 32px better than 36px?

**Comment by Pujan92** [comment]
As the issue persists due to the dynamic height calculation(37px for header & 32px for emojis), so 32px makes it of the same height and removes the inconsistency which happens due to the diff heights.

https://user-images.githubusercontent.com/14358475/207641095-15818c17-04f5-4f08-b94f-74f8f7c1196c.mp4


```diff
diff --git a/src/CONST.js b/src/CONST.js
index ff7e71b49..522cfaca9 100755
--- a/src/CONST.js
+++ b/src/CONST.js
@@ -495,11 +495,11 @@ const CONST = {
     ADD_PAYMENT_MENU_POSITION_X: 356,
     EMOJI_PICKER_SIZE: {
         WIDTH: 320,
-        HEIGHT: 400,
+        HEIGHT: 390,
     },
-    NON_NATIVE_EMOJI_PICKER_LIST_HEIGHT: 298,
+    NON_NATIVE_EMOJI_PICKER_LIST_HEIGHT: 288,
     EMOJI_PICKER_ITEM_HEIGHT: 32,
-    EMOJI_PICKER_HEADER_HEIGHT: 38,
+    EMOJI_PICKER_HEADER_HEIGHT: 32,
     COMPOSER_MAX_HEIGHT: 125,
     CHAT_FOOTER_MIN_HEIGHT: 65,
     CHAT_SKELETON_VIEW: {
diff --git a/src/components/EmojiPicker/EmojiPickerMenu/index.js b/src/components/EmojiPicker/EmojiPickerMenu/index.js
index 2d97a2443..fa90bdd25 100755
--- a/src/components/EmojiPicker/EmojiPickerMenu/index.js
+++ b/src/components/EmojiPicker/EmojiPickerMenu/index.js
@@ -508,6 +508,9 @@ class EmojiPickerMenu extends Component {
                             }
                             stickyHeaderIndices={this.state.headerIndices}
                             onScroll={e => this.currentScrollOffset = e.nativeEvent.contentOffset.y}
+                            getItemLayout={(data, index) => (
+                                {length: CONST.EMOJI_PICKER_ITEM_HEIGHT, offset: CONST.EMOJI_PICKER_ITEM_HEIGHT * index, index}
+                            )}
                         />
                     )}
                 <EmojiSkinToneList
 ```

**Comment by shawnborton** [comment]
Got it, thanks for clarifying. That looks good and makes sense to me!

**Comment by Pujan92** [comment]
@Santhosh-Sellavel can you plz check and review my last comment? It seems @shawnborton is fine with the height change of header to 32px.

**Comment by Santhosh-Sellavel** [comment]
Thanks Shawn, @Pujan92 I'll check test the behaviour and let you know!

**Comment by priyeshshah11** [comment]
> As the issue persists due to the dynamic height calculation(37px for header & 32px for emojis), so 32px makes it of the same height and removes the inconsistency which happens due to the diff heights.

@Pujan92 I don't think changing the heights of all these elements is really necessary here, this infact does more or same calculations than the above solutions which only calculates the heights once and then performs a map lookup which is just an O(1) operation

**Comment by Pujan92** [comment]
Sure @Santhosh-Sellavel , added comments on the changes line.

```diff
     EMOJI_PICKER_SIZE: {
         WIDTH: 320,
-        HEIGHT: 400,
+        HEIGHT: 390,  --    62px(TOP COMPOSER) + 228px(MIDDLE 9 ROWS WHICH IS 9*32px) + 40px(BOTTOM SKINTONE LIST)
     },
-    NON_NATIVE_EMOJI_PICKER_LIST_HEIGHT: 298,
+    NON_NATIVE_EMOJI_PICKER_LIST_HEIGHT: 288,   --   9 ROWS WHICH IS 9*32px
     EMOJI_PICKER_ITEM_HEIGHT: 32,
-    EMOJI_PICKER_HEADER_HEIGHT: 38,
+    EMOJI_PICKER_HEADER_HEIGHT: 32,  ---  NEW HEADER HEIGHT WHICH IS 32px


     emojiHeaderStyle: {
         backgroundColor: themeColors.componentBG,
         width: '100%',
-        ...spacing.pv3,
         fontFamily: fontFamily.EXP_NEUE_BOLD,
         fontWeight: fontWeightBold,
         color: themeColors.heading,
         fontSize: variables.fontSizeSmall,
+        height: 32,   ---  Giving fixed height to the header and aligning the item to the center vertically
+        display: 'flex',
+        alignItems: 'center',
     },

```

**Comment by Santhosh-Sellavel** [comment]
I am just confused why are we changing the HEIGHT of the picker 400 to 390 and 298 to 288?

**Comment by Pujan92** [comment]
> I am just confused why are we changing the HEIGHT of the picker 400 to 390 and 298 to 288?

To adjust and show only 9 rows of the flatlist to the view. If we keep it to 400 then the extra part will be visible partially.

eg.
<img width="342" alt="Screenshot 2022-12-15 at 6 09 39 AM" src="https://user-images.githubusercontent.com/14358475/207745433-7869a2ad-84e3-4177-aeab-d48857a995ae.png">

**Comment by Santhosh-Sellavel** [comment]
Thanks, @Pujan92!

@Gonals  @Pujan92's [proposal](https://github.com/Expensify/App/issues/12772#issuecomment-1351661286) Looks good let me know your thoughts!

**Comment by priyeshshah11** [comment]
@Santhosh-Sellavel @Gonals I still think the key solution here is to calculate the correct row height so actually there's no need to change all these sizes which is what @Ollyws & [my proposals](https://github.com/Expensify/App/issues/12772#issuecomment-1344497243) suggest.

**Comment by Santhosh-Sellavel** [comment]
> @Santhosh-Sellavel @Gonals I still think the key solution here is to calculate the correct row height which is what @Ollyws & [my proposals](https://github.com/Expensify/App/issues/12772#issuecomment-1344497243) suggest, so actually there's no need to change all these sizes.

Yes definitely if we are not okay with changing height, but we are good with it. Now this is much faster, also never a fan storing the size map a fairly simple component which has fixed heights for rows unlike chat view's list where size varies.

I'll leave it for @Gonals to  make the final call on this one, thanks!

**Comment by priyeshshah11** [comment]
> Yes definitely if we are not okay with changing height, but we are good with it. Now this is much faster, also never a fan storing the size map a fairly simple component which has fixed heights for rows unlike chat view's list where size varies.
> 
> I'll leave it for @Gonals to make the final call on this one, thanks!

I get that, if storing a size map is not ideal then [Proposal 1 from here](https://github.com/Expensify/App/issues/12772#issuecomment-1342754497) along with specifying the height in the style works too.
> Add height: CONST.EMOJI_PICKER_HEADER_HEIGHT, in emojiHeaderStyle in src/styles/styles.js

Happy to hear @Gonals thoughts on it.

**Comment by Santhosh-Sellavel** [comment]
> I get that, if storing a size map is not ideal then [Proposal 1 from here](https://github.com/Expensify/App/issues/12772#issuecomment-1342754497) along with specifying the height in the style works too.
> 
> > Add height: CONST.EMOJI_PICKER_HEADER_HEIGHT, in emojiHeaderStyle in src/styles/styles.js

I tried it earlier & now didn't work as expected, look selected emoji is not visible clearly.

<img width="388" alt="Screenshot 2022-12-16 at 5 12 37 AM" src="https://user-images.githubusercontent.com/85645967/207989860-08653226-8f3a-43b1-9a03-cab9beb81a42.png">

**Comment by priyeshshah11** [comment]
> I tried it earlier & now didn't work as expected, look selected emoji is not visible clearly.

yes you're right, I take that back. Thus, I still think Proposal 4 is ideal as I don't see anything wrong/inefficient with storing a size map for variable heights but that's a decision for you & @Gonals to make. I'll leave it to the experts, Thanks! 👍

**Comment by s77rt** [comment]
I think we still didn't find the root cause

**Comment by s77rt** [comment]
IMO @Pujan92 's [proposal](https://github.com/Expensify/App/issues/12772#issuecomment-1351661286) is the best solution so far. It works but we don't know why

**Comment by s77rt** [comment]
Quick Update: I have found the root cause.
The root cause is in RN or at least in RNW

Will post more details in few minutes...

**Comment by s77rt** [comment]
### Proposal (2)
```diff
diff --git a/node_modules/@expensify/react-native-web/dist/vendor/react-native/VirtualizedList/index.js b/node_modules/@expensify/react-native-web/dist/vendor/react-native/VirtualizedList/index.js
index ac40acc..0b8c778 100644
--- a/node_modules/@expensify/react-native-web/dist/vendor/react-native/VirtualizedList/index.js
+++ b/node_modules/@expensify/react-native-web/dist/vendor/react-native/VirtualizedList/index.js
@@ -1135,7 +1135,7 @@ class VirtualizedList extends React.PureComponent {
       if (!isVirtualizationDisabled && first > lastInitialIndex + 1) {
         var insertedStickySpacer = false;
 
-        if (stickyIndicesFromProps.size > 0) {
+        if (false && stickyIndicesFromProps.size > 0) {
           var stickyOffset = ListHeaderComponent ? 1 : 0; // See if there are any sticky headers in the virtualized space that we need to render.
 
           for (var ii = firstAfterInitial - 1; ii > lastInitialIndex; ii--) {
```
*I know the fix looks silly but I will explain*

### RCA
I have been digging in the code and I found that the issue is in RN (or at least it was)
This is the offending code https://github.com/Expensify/react-native/blob/main/Libraries/Lists/VirtualizedList.js#L715-L759
My first thought was that the issue is a miscalculation of `first` and `lastInitialIndex` or maybe somehow due to not taking the number of columns into consideration.

The weird part is that this code does nothing at all, is it trying to render something that is already rendered? So I was just about to open a PR on RN to get info about that code and fix it, But RN has already [refactored the VirtualizedList](https://github.com/facebook/react-native/commit/971599317b7bdf1152157206f9503a23ac8c4162) so there is a high chance that this has been fixed already.

However, RNW is still using the old version, I think we can open a PR / issue to "swap" the files, except that we have fork of RNW as well where we applied few changes on the VirtualizedList so switching to the new VirtualizedList won't be straightforward.

**Comment by s77rt** [comment]
@huzaifa-99 
> I believe its with the FlatList for web

VirtualizedList on RN but I think it was fixed already but not for RNW. Check the proposal above

**Comment by Santhosh-Sellavel** [comment]
@s77rt I don't get it changes make `if` block unreachable and even the issue still exists for me after making the changes.

**Comment by s77rt** [comment]
@Santhosh-Sellavel Sorry if it was unclear. As I stated in my first proposal this issue is related to sticky headers.
Apparently the issue is that RN/RNW is trying to render a header that is already rendered. In short, the block is useless. It was removed on the new refactored VirtualizedList.

Tests: Everything works fine and the tests passes.


After you apply the fix you have to clear the cache first: `rm -rf node_modules/.cache && npm run web`

**Comment by Santhosh-Sellavel** [comment]
> After you apply the fix you have to clear the cache first: `rm -rf node_modules/.cache && npm run web`

I tried still the issue exists.

Uploading Screen Recording 2022-12-18 at 2.48.59 AM.mov…

**Comment by s77rt** [comment]
@Santhosh-Sellavel 
> Apart from the RNW fix, we need to update the Header height in CONST to 37. Else we would get this partially hidden state.

Yes, that right.

**Comment by Santhosh-Sellavel** [comment]
@Gonals 

@s77rt's Proposal [here](https://github.com/Expensify/App/issues/12772#issuecomment-1356312398) looks good, along with updating header height.

**Comment by Pujan92** [comment]
@s77rt Great to see that you find the root cause. 🎉 

My only point is we are defining the height to the `NON_NATIVE_EMOJI_PICKER_LIST_HEIGHT` by initially considering (2 header `2*37px = 64px` and 7 emoji rows `7*32px = 224px` which sums to `298px`.
https://github.com/Expensify/App/blob/2483ad0a692b1392d7dc79f917edd44c5be46080/src/CONST.js#L501

Now when we scroll and on the view, if there is only 1 header in the view then in this case logically the height of the `NON_NATIVE_EMOJI_PICKER_LIST_HEIGHT` should be `293px(1 header 37px + 8 emoji rows 256px)`.
Changing the height of the list won't be a good option while scrolling and if we won't change that `5px` might cause partial visibility of the next line.

https://user-images.githubusercontent.com/14358475/208281857-7f5e758d-6280-46ca-a9c3-99cd824d0441.mov

**Comment by s77rt** [comment]
@Pujan92 I agree, however technically this is a different issue and should not be prioritised as it does not break the overall functionality. Fixing the `EMOJI_PICKER_HEADER_HEIGHT` should be enough. Besides there may be situations where dynamic height is required.

Ideally we should wait for RNW to update the `VirtualizedList` then update the Expensify's fork. Until then we can either apply a temporarily solution or disable the arrow navigation feature.

If the issue is on RN/RNW it must be fixed on RN/RNW

**Comment by Pujan92** [comment]
As we are mentioning the height/dynamic heights in the comments I thought it's better to point out the issue over here only. For fixed height, it seems to use `getItemLayout` as a better option for optimization purpose. I think for particular this use case this is a more permanent solution. As you stated for dynamic heights we should wait and depend on the sticky header fix from the RN.

`
getItemLayout is an optional optimization that allows skipping the measurement of dynamic content if you know the size (height or width) of items ahead of time. getItemLayout is efficient if you have fixed size items
`

**Comment by JmillsExpensify** [comment]
@Luke9389 It looks like @Gonals is now on vacation, and we're stuck getting agreement on the path forward. Can you please review the [latest comment](https://github.com/Expensify/App/issues/12772#issuecomment-1356478281) from @Santhosh-Sellavel so we can unblock this issue and keep it moving forward please? Thank you!

**Comment by Luke9389** [comment]
We've discussed internally and decided on the following course of action:

This proposal best: https://github.com/Expensify/App/issues/12772#issuecomment-1351661286
Though it does use `getItemLayout` which was first mentioned in @Ollyws's proposal here: https://github.com/Expensify/App/issues/12772#issuecomment-1340926774

For that reason we will move forward with @Pujan92's proposal, but award @Ollyws an additional $1000. 

I do appreciate that @s77rt found the root cause, but we do not have time to wait for RN to fix this on their end, and this never lead to a formal proposal to solve the issue.

**Comment by s77rt** [comment]
@Luke9389 I respect your decision but it feels as I'm being punished for following the rules and for trying to fix the issue upstream

**Comment by Luke9389** [comment]
Thanks for expressing that @s77rt. Before we continue with the issue, I want to be sure to give you a fair chance as well. 

So [this](https://github.com/Expensify/App/issues/12772#issuecomment-1356312398) is your proposal correct?

**Comment by s77rt** [comment]
@Luke9389 Yes, that's it, it solves the issue but still I would prefer to actually go with refactoring the VirtualizedList i.e. wait for RNW to adapt the new VirtualizedList.
Usually on such situations I'd expect the issue to be put on hold but if we can't wait for that I'd appreciate a compensation

**Comment by Luke9389** [comment]
I see. There are a few things to discuss here. 

> Yes, that's it, it solves the issue 

I'm curious to know why this is the case. After reading your explanation, I was still unsure.

> but still I would prefer to actually go with refactoring the VirtualizedList i.e. wait for RNW to adapt the new VirtualizedList.

While it would be ideal for RNW to adapt the new VirtualizedList, it wouldn't be a guarantee that we use it. The problem with this approach is that we are waiting for them to make a change, and _then we would need to enter a separate (and probably lengthy) dialogue to figure out if we should make the switch as well_. 

> Usually on such situations I'd expect the issue to be put on hold  

We would put this on hold if we agreed that we should wait for RNW to adapt VirtualizedList (and agreed on using it).

> I'd appreciate a compensation

The reason we have elected to give compensation to @Ollyws, is because the proposal we are going with includes some of their work. If the proposal we selected was building off of your ideas, then we would also compensate you. 

> it feels as I'm being punished for following the rules

This is the most important thing for us to reach a resolution about, because we want to build a community that operates with fair rules. Would you be willing to point out which rule you feel you're being punished for following? A link to the rule in documentation would be very much appreciated. Also, if you'd be willing to suggest an improvement to our rules, that would also be very helpful. 


I'll give you an opportunity to respond to these points before saying any more. Again, I just want to thank you for communicating with me about it. Our process is still undergoing change and we want to be sure everyone feels understood and listened to. 🙇

**Comment by s77rt** [comment]
@Luke9389 

> I'm curious to know how it solves the issue. After reading your explanation, I was still unsure.

The issue is that code will try to re-render the sticky headers that are already rendered and add more spaces that will effect the overall contentSIze and reposition the scroller. This part of code has been greatly refactored on the new VirtualizedList.

> While it would be ideal for RNW to adapt the new VirtualizedList, it wouldn't be a guarantee that we use it

I think we will have to use it the sooner or the latter.

> We would put this on hold if we agreed that we should wait for RNW to adapt VirtualizedList.

Understood.

> If the proposal we selected was building off of your ideas, then we would also compensate you

Understood

> Would you be willing to point out which rule you feel you're being punished for following?

Actually I can't find an explicit rule for what I said, but fixing the issues upstream is something that goes without saying, although I'm sure we were encouraged on other issues (not the rules) to fix issues upstream if the root cause is not on Expensify/App

> We want to be sure everyone feels understood and listened to

That's great, I appreciate your time for looking into my case.

**Comment by Luke9389** [comment]
Between my latest message and your response, I was made aware of [this thread](https://expensify.slack.com/archives/C01GTK53T8Q/p1671562714467819?thread_ts=1670868002.066289&cid=C01GTK53T8Q), which discusses our preference for updating packages upstream. 

Given all of this new information, it may be a good idea to change course and use VirtualizedList after all. But before doing that, I think we should get more details. 

Let's say that RNW adapted the refactored VirtualizedList. What changes would we then need to make? @s77rt, would you be willing to give a hypothetical proposal to help me determine what that would look like?

**Comment by s77rt** [comment]
What changes would we then need to make?
1. To fix this issue?
Hopefully nothing, but if the issue still persist, we will have to create a PR in RN. Although the refactored VirtualizedList looks promising we can't be sure as we can't test it unless RNW adopts it.

2. In general
We will have to verify that those commits https://github.com/Expensify/react-native-web/commits/master/packages/react-native-web/src/vendor/react-native/VirtualizedList/index.js can still be applied elegantly to fix whatever they were designed to fix

The point 2 is unavoidable, if at any time RNW adopts the new VirtualizedList we will eventually do the same


Hypothetical proposal: (excluding point 2 because it's not directly linked to this issue and it's something we will do anyway, I suppose)
- Best case scenario: do nothing
- Worst case scenario: fix the persisting issue in RN

**Comment by Luke9389** [comment]
Is there any way for us to test whether using the updated VirtualizedList will work?

**Comment by s77rt** [comment]
Maybe if we add the arrow navigation feature to native and update our forked RN to use the updated VirtualizedList, then yes it would be possible

**Comment by Luke9389** [comment]
Sorry which arrow navigation feature are you referring to? 

Ultimately, what I'm really looking for is a clear plan for exactly how you think we should move forward. We could apply the above solution and close this issue soon, but if you can convince us that you have a better alternative, please spell it out in one detailed comment with a clear proposal and steps to follow.

This is your chance to change our course!

**Comment by s77rt** [comment]
> Sorry which arrow navigation feature are you referring to?

There is only one arrow navigation feature, it's what this issue is about. You can navigate through emojis using your keyboard, this is not available on native, but if we can replicate this behaviour on native then we can test whether the updated VirtualizedList will work.

Sometimes the ideal solution is too much time consuming that makes it feels less of an ideal.

Maybe it would be a good call to apply a temp solution or disable arrow navigation until:
1. RNW adapts the refactored VirtualizedList https://github.com/necolas/react-native-web/issues/2453
2. Expensify's RNW fork adapts the refactored VirtualizedList ( and re-apply https://github.com/Expensify/react-native-web/commits/master/packages/react-native-web/src/vendor/react-native/VirtualizedList/index.js )

Only after point 1 and point 2 are done, we can actually fix this issue properly.

___

Summary:
PLAN A:
1. Wait for RNW to adapt the refactored VirtualizedList
2. We should also adapt the refactored VirtualizedList
3. Issue is fixed.

PLAN B:
1. Wait for RNW to adapt the refactored VirtualizedList
2. We should also adapt the refactored VirtualizedList
3. Issue is still persist, re-open this GH issue for proposals that aim to fix RN's bug
4. Once the bug is identified, create a PR on RN and possibly on RNW too
5. Issue is fixed.

**Comment by priyeshshah11** [comment]
> This proposal best: [#12772 (comment)](https://github.com/Expensify/App/issues/12772#issuecomment-1351661286) Though it does use `getItemLayout` which was first mentioned in @Ollyws's proposal here: [#12772 (comment)](https://github.com/Expensify/App/issues/12772#issuecomment-1340926774)

@Luke9389 Have you tried this https://github.com/Expensify/App/issues/12772#issuecomment-1344497243 solution? I feel this https://github.com/Expensify/App/issues/12772#issuecomment-1351661286 solution avoids the problem rather than fixing it.

**Comment by Pujan92** [comment]
In this case, where we have the fixed height, I think `getItemLayout` will be the preferable approach based on the react native [doc](https://reactnative.dev/docs/flatlist#getitemlayout) too. @s77rt as per my understanding the RN fix which you have suggested helps for dynamic heights flatlist which doesn't seem the case here.

**Comment by Pujan92** [comment]
In this case, where we have the fixed height, I think `getItemLayout` will be the preferable approach based on the react native [doc](https://reactnative.dev/docs/flatlist#getitemlayout) too. @s77rt as per my understanding the RN fix which you have suggested helps for dynamic heights flatlist which doesn't seem the case here.

**Comment by Pujan92** [comment]
> @Luke9389 Have you tried this [#12772 (comment)](https://github.com/Expensify/App/issues/12772#issuecomment-1344497243) solution? I feel this [#12772 (comment)](https://github.com/Expensify/App/issues/12772#issuecomment-1351661286) solution avoids the problem rather than fixing it.

@priyeshshah11 I have explained there and [here](https://github.com/Expensify/App/issues/12772#issuecomment-1356670553) why the same height is required for good of the application perspective and for the same get approval from the design team(@shawnborton) too. So I don't think this solution is avoiding the problem from any direction.

**Comment by priyeshshah11** [comment]
> @priyeshshah11 I have explained there and [here](https://github.com/Expensify/App/issues/12772#issuecomment-1356670553) why the same height is required for good of the application perspective and for the same get approval from the design team(@shawnborton) too. So I don't think this solution is avoiding the problem from any direction.

@Pujan92 I understand you're point but IMO changing the header height to be the same as the item height doesn't actually fix the actual bug, it avoids it. I understand that the design team is ok to change the height in this case but sometimes that might not be possible so IMO changing the design to fix the problem should be the absolute last resort where we can't find any good enough solutions. 

Also in regards to the efficiency of the application, `getItemLayout` is equally efficient with variable heights when you know the exact heights in advance, which we do in this case.

**Comment by Luke9389** [comment]
Ok so I think we should go with the plan I posted here: https://github.com/Expensify/App/issues/12772#issuecomment-1360227084

> I understand that the design team is ok to change the height in this case but sometimes that might not be possible so IMO changing the design to fix the problem should be the absolute last resort where we can't find any good enough solutions.

@priyeshshah11 I appreciate that you want to build a more general solution that could be used elsewhere, but the reality is that this is a small issue that we aren't seeing anywhere else on the app. It just doesn't need [this level of complexity](https://github.com/Expensify/App/issues/12772#issuecomment-1344497243). I do think you've got great ideas, and you should continue to think this way for other issues, especially larger ones. 

@s77rt, you did a good job of laying out your point, but at this time we don't want to wait for RNW and we don't want to disable arrow navigation on the emoji picker. You mentioned this

> Maybe it would be a good call to apply a temp solution 

and I think in this case, that's what we'll do. Thanks for all the time and energy you put toward this discussion.


With all that said, @Pujan92 you've got the 🟢 green light 🟢 to start your PR. @Santhosh-Sellavel and I will see you there.

**Comment by Pujan92** [comment]
Thanks, applied for the job and PR is ready for review.

**Comment by JmillsExpensify** [comment]
It looks like only @Santhosh-Sellavel is assigned for review. @Luke9389 should you also be on the PR, or do we need to pull bear so that others are assigned? Let's help @Pujan92 get this merged in three business days!

**Comment by JmillsExpensify** [comment]
Actually...something weird is going on? Looks like @Gonals was chosen but I'm pretty sure he's OOO now and also not assigned.

**Comment by Pujan92** [comment]
> Actually...something weird is going on? Looks like @Gonals was chosen but I'm pretty sure he's OOO now and also not assigned.

I think it is my mistake and happens due to the incorrect way of linking the issue in the PR. Updated the PR.

**Comment by JmillsExpensify** [comment]
Looks like the linked PR is ready to merge. Left a comment for @Gonals to confirm in the PR.

**Comment by JmillsExpensify** [comment]
Today is payday on this issue, so I'm circling back to close the loop on a payment summary. Looking at the PR, I'm going to go ahead and award C/C+ the bonus for urgency. Any delay in merging the PR was caused on our side due to the holidays. That means that payouts will look like this:

 - Issue reporter: @parasharrajat - $250
 - Contributor: @Pujan92 - $6,000 (incl 50% bonus)
 - Contributor+: @Santhosh-Sellavel - $6,000 (incl 50% bonus). 

Upwork job [is here](https://www.upwork.com/jobs/~011a004a1208ec7fec). An offer has been extended to @Pujan92, @Santhosh-Sellavel has been invited, as has @parasharrajat. Thanks!

**Comment by JmillsExpensify** [comment]
Ah thank you! Can you please apply to the Upwork job? I'll issue an offer once you apply.

**Comment by JmillsExpensify** [comment]
@Pujan92 and @Ollyws have been paid. @Santhosh-Sellavel was sent an offer (I'll pay out the bonus when payment is issued). Still waiting on @parasharrajat to accept the invite.

**Comment by JmillsExpensify** [comment]
Then @Santhosh-Sellavel @Luke9389 Can either of you kick-off and complete your items on the BZ checklist? I've already created an issue for regression testing in the Expensify/Expensify repo.

**Comment by Luke9389** [comment]
I'm not sure if this is a regression. I think this behavior has always been here and we just recently noticed it. If that's the case, should we check the "The PR that introduced the bug has been identified." box? 

I'll dig in to see if I can find an offending PR but I'd still like to know what to do in the case that there isn't one.

**Comment by JmillsExpensify** [comment]
If there isn't a regression, then all good. We should be able to cover this moving forward with the regression test we'll add moving forward.

**Comment by Luke9389** [comment]
Ok, I didn't see anything that indicates this is a regression, so I checked us off.

**Comment by JmillsExpensify** [comment]
@Santhosh-Sellavel has been paid out, and all items are successfully checked off in the BZ checklist. Closing this issue as a result.
